src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs

```diff
@@ -5,19 +5,193 @@
+            : base(assertIsForeground: false)
+        {
+            _serviceProvider = serviceProvider;
+        }
+
+        public bool IsSupported(Workspace workspace)
             => workspace is VisualStudioWorkspaceImpl;
 
-        public string GetStorageLocation(Solution solution)
+        public string TryGetStorageLocation(SolutionId solutionId)
+        {
+            lock (_gate)
+            {
+                if (solutionId == _currentSolutionId)
+                {
+                    return _currentWorkingFolderPath;
+                }
+            }
+
+            return null;
+        }
+
+        internal void UpdateForVisualStudioWorkspace(Workspace visualStudioWorkspace)
+        {
+            AssertIsForeground();
+
+            lock (_gate)
+            {
+                if (visualStudioWorkspace.CurrentSolution.Id == _currentSolutionId && _currentWorkingFolderPath != null)
+                {
+                    return;
+                }
+
+                var solution = _serviceProvider.GetService<SVsSolution, IVsSolution>();
+
+                if (!_solutionEventsAdvised)
+                {
+                    solution.AdviseSolutionEvents(new EventSink(this, visualStudioWorkspace), out uint cookie);
+                    _solutionEventsAdvised = true;
+                }
+
+                try
+                {
+                    var solutionWorkingFolder = (IVsSolutionWorkingFolders)solution;
+                    solutionWorkingFolder.GetFolder(
+                        (uint)__SolutionWorkingFolder.SlnWF_StatePersistence, Guid.Empty, fVersionSpecific: true, fEnsureCreated: true,
+                        pfIsTemporary: out var temporary, pszBstrFullPath: out var workingFolderPath);
+
+                    if (!temporary && !string.IsNullOrWhiteSpace(workingFolderPath))
+                    {
+                        OnWorkingFolderChanging_NoLock(
+                            new PersistentStorageLocationChangingEventArgs(
+                                visualStudioWorkspace.CurrentSolution.Id,
+                                workingFolderPath,
+                                mustUseNewStorageLocationImmediately: false));
+                    }
+                }
+                catch
+                {
+                    // don't crash just because solution having problem getting working folder information
+                }
+            }
+        }
+
+        private void OnWorkingFolderChanging_NoLock(PersistentStorageLocationChangingEventArgs eventArgs)
+        {
+            AssertIsForeground();
+
+            StorageLocationChanging?.Invoke(this, eventArgs);
+
+            _currentSolutionId = eventArgs.SolutionId;
+            _currentWorkingFolderPath = eventArgs.NewStorageLocation;
+        }
+
+        private void ProcessChangeToIVsSolutionChange(Workspace visualStudioWorkspace)
+        {
+            lock (_gate)
```

Comment:
threading assert?

Reply:
i would think this could only be called from the UI thread given htat it's called from a VS api sink below..  If that's the case, do we need the lock, or is all of this UI thread affinitized?

Reply:
All mutation is affinitized, but TryGetStorageLocation() looks at two fields while reading so we don't want any strange half-written states going on there.